lib/app_profiler/middleware/upload_action.rb

```diff
@@ -16,8 +21,54 @@ def call(profile, response: nil, autoredirect: nil)
           )
         end
 
+        def enqueue_upload(profile)
+          mutex.synchronize do
+            @queue ||= init_queue
+            begin
+              @queue.push(profile, true) # non-blocking push, raises ThreadError if queue is full
+            rescue ThreadError
+              AppProfiler.logger.info("[AppProfiler] upload queue is full, profile discarded")
+            end
+          end
+        end
+
+        def start_process_queue_thread
+          @process_queue_thread ||= Thread.new do
+            loop do
+              process_queue
+              sleep(AppProfiler.upload_queue_interval_secs)
+            end
+          end
+          @process_queue_thread.priority = -1 # low priority
```

Comment:
Not exactly sure if this is needed -- but the idea was to give a low priority from the main thread. https://ruby-doc.org/core-2.5.0/Thread.html#method-i-priority